client/src/lib/games/ddr.tsx

```diff
@@ -0,0 +1,154 @@
+import { GetEnumColour } from "lib/game-implementations";
+import React from "react";
+import RatingCell from "components/tables/cells/RatingCell";
+import LampCell from "../../components/tables/cells/LampCell";
+import DDRScoreCell from "../../components/tables/cells/DDRScoreCell";
+import { bg, bgc } from "./_util";
+
+const DDR_ENUM_COLOURS: GPTClientImplementation<GPTStrings["ddr"]>["enumColours"] = {
+	grade: {
+		E: COLOUR_SET.gray,
+		D: COLOUR_SET.paleBlue,
+		"D+": COLOUR_SET.paleBlue,
+		"C-": COLOUR_SET.purple,
+		C: COLOUR_SET.purple,
+		"C+": COLOUR_SET.purple,
+		"B-": COLOUR_SET.blue,
+		B: COLOUR_SET.blue,
+		"B+": COLOUR_SET.blue,
+		"A-": COLOUR_SET.gold,
+		A: COLOUR_SET.gold,
+		"A+": COLOUR_SET.gold,
+		"AA-": COLOUR_SET.gold,
+		AA: COLOUR_SET.gold,
+		"AA+": COLOUR_SET.gold,
+		AAA: COLOUR_SET.teal,
+	},
+	lamp: {
+		FAILED: COLOUR_SET.maroon,
+		ASSIST: COLOUR_SET.purple,
+		CLEAR: COLOUR_SET.blue,
+		"FULL COMBO": COLOUR_SET.vibrantBlue,
+		"GREAT FULL COMBO": COLOUR_SET.vibrantGreen,
+		"PERFECT FULL COMBO": COLOUR_SET.gold,
+		"MARVELOUS FULL COMBO": COLOUR_SET.pink,
+		LIFE4: COLOUR_SET.vibrantRed,
+	},
+};
+
+const DDR_DIFF_COLOURS: GPTClientImplementation<GPTStrings["ddr"]>["difficultyColours"] = {
+	BEGINNER: COLOUR_SET.blue,
+	BASIC: COLOUR_SET.paleGreen,
+	DIFFICULT: COLOUR_SET.red,
+	EXPERT: COLOUR_SET.vibrantYellow,
+	CHALLENGE: COLOUR_SET.purple,
+};
+
+const DDR_HEADERS: GPTClientImplementation<"ddr:SP" | "ddr:DP">["scoreHeaders"] = [
+	["Score", "Score", NumericSOV((x) => x.scoreData.percent)],
+	["Lamp", "Lamp", NumericSOV((x) => x.scoreData.enumIndexes.lamp)],
+];
+
+const DDR_COLOURS: GPTClientImplementation<"ddr:SP" | "ddr:DP">["classColours"] = {
+	flare: {
+		NONE: bg("gray"),
+		"NONE+": bg("gray"),
+		"NONE++": bg("gray"),
+		"NONE+++": bg("gray"),
+		MERCURY: bg("blue"),
+		"MERCURY+": bg("blue"),
+		"MERCURY++": bg("blue"),
+		"MERCURY+++": bg("blue"),
+		VENUS: bgc("yellow", "black"),
+		"VENUS+": bgc("yellow", "black"),
+		"VENUS++": bgc("yellow", "black"),
+		"VENUS+++": bgc("yellow", "black"),
+		EARTH: bg("forestgreen"),
+		"EARTH+": bg("forestgreen"),
+		"EARTH++": bg("forestgreen"),
+		"EARTH+++": bg("forestgreen"),
+		MARS: bg("red"),
+		"MARS+": bg("red"),
+		"MARS++": bg("red"),
+		"MARS+++": bg("red"),
+		JUPITER: bg("darkgreen"),
+		"JUPITER+": bg("darkgreen"),
+		"JUPITER++": bg("darkgreen"),
+		"JUPITER+++": bg("darkgreen"),
+		SATURN: bg("purple"),
+		"SATURN+": bg("purple"),
+		"SATURN++": bg("purple"),
+		"SATURN+++": bg("purple"),
+		URANUS: bgc("powderblue", "black"),
+		"URANUS+": bgc("powderblue", "black"),
+		"URANUS++": bgc("powderblue", "black"),
+		"URANUS+++": bgc("powderblue", "black"),
+		NEPTUNE: bg("darkslateblue"),
+		"NEPTUNE+": bg("darkslateblue"),
+		"NEPTUNE++": bg("darkslateblue"),
+		"NEPTUNE+++": bg("darkslateblue"),
+		SUN: bg("orange"),
+		"SUN+": bg("orange"),
+		"SUN++": bg("orange"),
+		"SUN+++": bg("orange"),
+		WORLD: bg("black"),
```

Comment:
ouch, not fun to write or display...

Reply:
Do you want any change in the colours used or is it just a tease about the 41 ranks that all need a colour ? 😋

Reply:
just a tease sorry! github doesn't make it easy to distinguish "change needed here" and "this is just a comment on the PR".